uvtools/dspec.py

```diff
@@ -1235,9 +1241,62 @@ def delay_filter_leastsq(data, flags, sigma, nmax, add_noise=False, freq_units =
     return mdl_array, cn_array, inp_data
 
 
-def sinc_downweight_mat_inv(nchan, df, filter_centers, filter_widths,
-                            filter_factors, cache = {}, wrap = False, wrap_interval=1,
-                            nwraps = 1000, no_regularization = False):
+def delay_interpolation_matrix(nchan, ndelay, wgts, fundamental_period=None, cache={}, taper='none', return_diagnostics=False):
+    '''
+    Computes a foreground interpolation matrix that, when applied to data,
+    interpolates over flagged channels with delays between -ndelays//2 and ndelays//2.
+
+    Parameters
+    ----------
+    nchan: integer
+        Number of frequency channels to interpolate over
+    ndelay,
+        number of delays to use in interpolation
+    wgts: float array
+        wgts to be applied to each frequency channel.
+    fundamental_period: float, optional
+        the fundamental period of reconstructed delays. Default: nchan
+    cache: dict, optional
+        optional cache holding pre-computed matrices
+    taper: string, optional
+        use a taper to fit.
+    Returns
+    ----------
+    (nchan, nchan) numpy array
+        that can be used to interpolate over channel gaps.
+    '''
+    assert len(wgts) == nchan, "nchan must equal length of wgts"
+    matkey = (nchan, ndelay) + tuple(wgts)
+    assert np.sum((np.abs(wgts) > 0.).astype(float)) >= ndelay, "number of unflagged channels must be greater then or equal to number of delays"
+    if fundamental_period is None:
+        fundamental_period = nchan #this tends to give well conditioned matrices. 
+    if not matkey in cache or return_diagnostics:
+        f, d = np.meshgrid(np.arange(nchan)-nchan/2, np.arange(-ndelay,ndelay), indexing='ij')
+        #f = np.asarray([[chan for m in range(ndelay)] for chan in np.arange(-nchan//2,nchan//2)])
+        #d = np.asarray([np.arange(-ndelay//2, ndelay//2) for m in range(nchan)])
+        d = d / fundamental_period
+        a_mat = np.exp(2j * d * f * np.pi) / fundamental_period
+        wmat = np.diag(wgts * gen_window(taper, nchan)).astype(complex)
+        cmat=np.dot(a_mat.T, (a_mat.T * wgts).T)
+        if np.linalg.cond(cmat)>=1e9:
+            print('Warning!!!!: Poorly conditioned matrix! Your linear inpainting IS WRONG!'
```

Comment:
Rather than a print statement, this should be raised using the `warnings` module. (E.g., `warnings.warn("Poorly conditioned matrix! ...`)

Reply:
I've changed this to a warning.